File: apps/web_companion/src/state/friend_store.ts
import { createStore, type Store } from './store';

/**
 * The social service has no friend-listing API yet (GET_FRIEND_LIST 3007 is
 * unimplemented), so the client keeps its own roster: everyone we sent a
 * request to or accepted one from. Persisted per user in localStorage as a
 * stopgap — the authoritative list arrives with the P1 aggregation plane.
 */
export interface FriendState {
  /** Accepted friends (user ids, sorted). */
  friends: string[];
  /** Requests we received and have not answered. */
  pendingIn: Array<{ requestId: string; fromUserId: string }>;
  /** Requests we sent that nobody has answered yet. */
  pendingOut: string[];
}

export const createFriendStore = (initial: FriendState = { friends: [], pendingIn: [], pendingOut: [] }): Store<FriendState> =>
  createStore<FriendState>(initial);

export function addFriend(store: Store<FriendState>, userId: string): void {
  store.set((prev) =>
    prev.friends.includes(userId)
      ? prev
      : { ...prev, friends: [...prev.friends, userId].sort() },
  );
}

export function removeFriend(store: Store<FriendState>, userId: string): void {
  store.set((prev) =>
    prev.friends.includes(userId) ? { ...prev, friends: prev.friends.filter((f) => f !== userId) } : prev,
  );
}

export function addPendingOut(store: Store<FriendState>, userId: string): void {
  store.set((prev) =>
    prev.pendingOut.includes(userId) || prev.friends.includes(userId)
      ? prev
      : { ...prev, pendingOut: [...prev.pendingOut, userId].sort() },
  );
}

export function addPendingIn(store: Store<FriendState>, requestId: string, fromUserId: string): void {
  store.set((prev) =>
    prev.pendingIn.some((r) => r.requestId === requestId)
      ? prev
      : { ...prev, pendingIn: [...prev.pendingIn, { requestId, fromUserId }] },
  );
}

export function resolvePending(store: Store<FriendState>, requestId: string): void {
  store.set((prev) => ({
    ...prev,
    pendingIn: prev.pendingIn.filter((r) => r.requestId !== requestId),
  }));
}

export const fromUserIdOf = (
  state: FriendState,
  requestId: string,
): string | undefined => state.pendingIn.find((r) => r.requestId === requestId)?.fromUserId;

const friendStorageKey = (userId: string): string => `chirp.friends.${userId}`;

/** Best-effort load; storage may be unavailable (private mode) or corrupt. */
export function loadFriendState(userId: string): FriendState {
  const empty: FriendState = { friends: [], pendingIn: [], pendingOut: [] };
  try {
    const raw = window.localStorage.getItem(friendStorageKey(userId));
    if (!raw) return empty;
    const parsed = JSON.parse(raw) as Partial<FriendState>;
    return {
      friends: parsed.friends ?? [],
      pendingIn: parsed.pendingIn ?? [],
      pendingOut: parsed.pendingOut ?? [],
    };
  } catch {
    return empty;
  }
}

export function saveFriendState(userId: string, state: FriendState): void {
  try {
    window.localStorage.setItem(friendStorageKey(userId), JSON.stringify(state));
  } catch {
    // Non-fatal: the roster is a local stopgap anyway.
  }
}

/** Subscribe so every mutation lands in localStorage (fire and forget). */
export function persistFriendStore(store: Store<FriendState>, userId: string): () => void {
  const save = (): void => saveFriendState(userId, store.get());
  save();
  return store.subscribe(save);
}

File: apps/mobile_companion/lib/state/friend_store.dart
import 'store.dart';

class PendingRequest {
  const PendingRequest({required this.requestId, required this.fromUserId});

  final String requestId;
  final String fromUserId;
}

/// Client-side mirror of the social plane's roster. The backend is
/// authoritative: login pulls the friend list and the incoming request queue
/// (GET_FRIEND_LIST / GET_PENDING_REQUESTS), and notifications keep the mirror
/// current. Only pendingOut lives purely client-side — the service has no
/// outgoing-request query yet, so an app restart drops that list.
class FriendState {
  const FriendState({
    required this.friends,
    required this.pendingIn,
    required this.pendingOut,
  });

  /// Accepted friends (user ids, sorted).
  final List<String> friends;

  /// Requests we received and have not answered.
  final List<PendingRequest> pendingIn;

  /// Requests we sent that nobody has answered yet.
  final List<String> pendingOut;
}

Store<FriendState> createFriendStore() => Store<FriendState>(const FriendState(
      friends: [],
      pendingIn: [],
      pendingOut: [],
    ));

void replaceFriends(Store<FriendState> store, List<String> ids) {
  final sorted = List.of(ids)..sort();
  store.update((prev) {
    final same = prev.friends.length == sorted.length &&
        Iterable<int>.generate(sorted.length)
            .every((i) => prev.friends[i] == sorted[i]);
    return same
        ? prev
        : FriendState(
            friends: sorted,
            pendingIn: prev.pendingIn,
            pendingOut: prev.pendingOut,
          );
  });
}

void replacePendingIn(Store<FriendState> store, List<PendingRequest> requests) {
  store.update((prev) => FriendState(
        friends: prev.friends,
        pendingIn: requests,
        pendingOut: prev.pendingOut,
      ));
}

void addFriend(Store<FriendState> store, String userId) {
  store.update((prev) => prev.friends.contains(userId)
      ? prev
      : FriendState(
          friends: [...prev.friends, userId]..sort(),
          pendingIn: prev.pendingIn,
          pendingOut: prev.pendingOut,
        ));
}

void removeFriendStore(Store<FriendState> store, String userId) {
  store.update((prev) => prev.friends.contains(userId)
      ? FriendState(
          friends: prev.friends.where((f) => f != userId).toList(),
          pendingIn: prev.pendingIn,
          pendingOut: prev.pendingOut,
        )
      : prev);
}

void addPendingOut(Store<FriendState> store, String userId) {
  store.update((prev) =>
      prev.pendingOut.contains(userId) || prev.friends.contains(userId)
          ? prev
          : FriendState(
              friends: prev.friends,
              pendingIn: prev.pendingIn,
              pendingOut: [...prev.pendingOut, userId]..sort(),
            ));
}

void addPendingIn(
    Store<FriendState> store, String requestId, String fromUserId) {
  store.update((prev) => prev.pendingIn.any((r) => r.requestId == requestId)
      ? prev
      : FriendState(
          friends: prev.friends,
          pendingIn: [
            ...prev.pendingIn,
            PendingRequest(requestId: requestId, fromUserId: fromUserId),
          ],
          pendingOut: prev.pendingOut,
        ));
}

void resolvePending(Store<FriendState> store, String requestId) {
  store.update((prev) => FriendState(
        friends: prev.friends,
        pendingIn:
            prev.pendingIn.where((r) => r.requestId != requestId).toList(),
        pendingOut: prev.pendingOut,
      ));
}

String? fromUserIdOf(FriendState state, String requestId) {
  for (final request in state.pendingIn) {
    if (request.requestId == requestId) return request.fromUserId;
  }
  return null;
}
